### PR TITLE
refactor(2/12): add pipeline event types, parsers, and event builders

### PR DESCRIPTION
**Pull request by cameroncooke**
## Summary

This is **PR 2 of 12** in a stacked PR series that decouples the rendering pipeline from MCP transport. Depends on PR 1 (logging removal). This PR is additive -- it introduces new modules with no changes to existing code.

Introduces the foundational type system and parsing infrastructure for the new rendering pipeline. The core idea: tools communicate their results through structured `PipelineEvent` objects rather than constructing MCP `ToolResponse` content directly. This decouples what a tool wants to say from how it gets rendered.

### Architecture

The pipeline event model has three layers:

1. **Event types** (`src/types/pipeline-events.ts`): A discriminated union of all event kinds -- `header`, `status-line`, `detail-tree`, `diagnostic-summary`, `next-steps`, `progress`, etc. Each event is a plain data object with no rendering logic.

2. **Parsers** that produce events from xcodebuild/Swift Testing output:
   - `xcodebuild-event-parser.ts`: Converts raw xcodebuild stdout lines into structured events (build progress, warnings, errors, test results)
   - `xcodebuild-line-parsers.ts`: Low-level regex-based line classifiers for xcodebuild output
   - `swift-testing-event-parser.ts`: Handles Swift Testing's JSON event stream format
   - `swift-testing-line-parsers.ts`: Line classifiers for Swift Testing console output
   - `xcodebuild-run-state.ts`: Stateful tracker that deduplicates and orders events during a build/test run
   - `xcodebuild-error-utils.ts`: Error extraction helpers

3. **Event builders** (`tool-event-builders.ts`): Factory functions (`header()`, `statusLine()`, `detailTree()`, etc.) that construct properly-typed events. These are what tool handlers call.

### Why this comes first

Every subsequent PR in the stack depends on these types. The parsers are used by the xcodebuild pipeline (PR 4), the event builders are used by every tool handler (PRs 6-9), and the event types flow through the rendering engine (PR 3) and runtime invoker (PR 5).

## Stack navigation

- PR 1/12: Remove deprecated logging tools
- **PR 2/12** (this PR): Pipeline event types, parsers, and event builders
- PR 3/12: Rendering engine and output formatting
- PR 4/12: Build/test utility extraction, platform steps, xcodebuild pipeline
- PR 5/12: Runtime handler contract and tool invoker
- PR 6-9/12: Tool migrations (simulator, device/macOS, UI automation, remaining)
- PR 10/12: CLI, daemon, and MCP server boundaries
- PR 11-12/12: Config, docs, snapshot tests

## Test plan

- [ ] `npx vitest run` passes -- new test files for each parser module
- [ ] Event type discriminated union is exhaustive (verified by TypeScript compiler)
- [ ] Parser tests cover edge cases (malformed lines, empty input, interleaved output)

### CODE REVIEW

File: src/utils/swift-testing-event-parser.ts [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+    }
+    onEvent({
+      type: 'test-failure',
+      timestamp: now(),
+      operation: 'TEST',
+      test: lastIssueDiagnostic.testName,
+      message: lastIssueDiagnostic.message,
+      location: lastIssueDiagnostic.location,
+    });
+    lastIssueDiagnostic = null;
```
[the remark this answers is not in the document]
Comment (cameroncooke) [reply src/utils/swift-testing-event-parser.ts:60]: Fixed in latest push. Added `suiteName` to `lastIssueDiagnostic` type and propagated it through to the emitted `test-failure` event.

File: src/utils/xcodebuild-run-state.ts [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+    ): XcodebuildRunState {
+      state.finalStatus = succeeded ? 'SUCCEEDED' : 'FAILED';
+      state.wallClockDurationMs = durationMs ?? null;
+
+      if (options?.emitSummary !== false) {
+        const reconciledFailedTests = Math.max(state.failedTests, state.testFailures.length);
+        const reconciledPassedTests = Math.max(
+          0,
+          state.completedTests - state.failedTests - state.skippedTests,
+        );
```
[the remark this answers is not in the document]
Comment (cameroncooke) [reply src/utils/xcodebuild-run-state.ts:203]: Investigated: this is not a bug. `reconciledPassedTests` correctly uses `state.failedTests` (the progress-counter value) for the passed-test arithmetic. `reconciledFailedTests` is only used for the summary display to account for diagnostic-count discrepancies. The two serve different purposes.

**Comment by cameroncooke** [comment]
* **#330** <a href="https://app.graphite.com/github/pr/getsentry/XcodeBuildMCP/330?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#329** <a href="https://app.graphite.com/github/pr/getsentry/XcodeBuildMCP/329?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#328** <a href="https://app.graphite.com/github/pr/getsentry/XcodeBuildMCP/328?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#327** <a href="https://app.graphite.com/github/pr/getsentry/XcodeBuildMCP/327?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#326** <a href="https://app.graphite.com/github/pr/getsentry/XcodeBuildMCP/326?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#325** <a href="https://app.graphite.com/github/pr/getsentry/XcodeBuildMCP/325?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#324** <a href="https://app.graphite.com/github/pr/getsentry/XcodeBuildMCP/324?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#323** <a href="https://app.graphite.com/github/pr/getsentry/XcodeBuildMCP/323?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#322** <a href="https://app.graphite.com/github/pr/getsentry/XcodeBuildMCP/322?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#321** <a href="https://app.graphite.com/github/pr/getsentry/XcodeBuildMCP/321?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#320** <a href="https://app.graphite.com/github/pr/getsentry/XcodeBuildMCP/320?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/getsentry/XcodeBuildMCP/320?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#319** <a href="https://app.graphite.com/github/pr/getsentry/XcodeBuildMCP/319?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: src/utils/xcodebuild-line-parsers.ts [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+export function parseRawTestName(rawName: string): { suiteName?: string; testName: string } {
+  const objcMatch = rawName.match(/^-\[(.+?)\s+(.+)\]$/u);
+  if (objcMatch) {
+    return { suiteName: normalizeSuiteName(objcMatch[1]), testName: objcMatch[2] };
+  }
+
+  const slashParts = rawName.split('/').filter(Boolean);
+  if (slashParts.length >= 3) {
+    return { suiteName: `${slashParts[0]}/${slashParts[1]}`, testName: slashParts[2] };
```
[the remark this answers is not in the document]
Comment (cameroncooke) [reply src/utils/xcodebuild-line-parsers.ts]: Fixed in 904c110b. Changed `slashParts[0]`/`[1]`/`[2]` to `slashParts.slice(0, -1).join('/')` and `slashParts.at(-1)!` so arbitrarily nested Swift Testing suite names are preserved.

File: src/utils/swift-testing-event-parser.ts [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+      return;
+    }
+
+    // Swift Testing run summary
+    const stSummary = parseSwiftTestingRunSummary(line);
+    if (stSummary) {
+      completedCount = stSummary.executed;
+      failedCount = stSummary.failed;
+      emitTestProgress();
+      return;
```
[the remark this answers is not in the document]
Comment (cameroncooke) [reply src/utils/swift-testing-event-parser.ts:144]: Not a bug. The Swift Testing summary format doesn't report a skipped count, so `skippedCount` is only available from incremental result lines. The incremental value is the only source of truth here -- resetting it would lose data. `completedCount` and `failedCount` are overwritten because the summary provides authoritative values for those; no equivalent exists for skipped.

File: src/utils/xcodebuild-run-state.ts [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+        case 'test-failure': {
+          const key = normalizeTestFailureKey(event);
+          if (seenDiagnostics.has(key)) {
+            return;
+          }
+          seenDiagnostics.add(key);
+          state.testFailures.push(event);
+          accept(event);
+          break;
+        }
```
[the remark this answers is not in the document]
Comment (cameroncooke) [reply src/utils/xcodebuild-run-state.ts:150]: Not a bug. The two key formats are structurally distinct and can't collide. Compiler diagnostics use `normalizeDiagnosticKey` which preserves the full path (`/full/path/file.swift:42:10|message`), while test failures use `normalizeTestFailureKey` which strips to just `filename:line:col|message` or the three-part `suite|test|message` format. These namespaces don't overlap.

File: src/utils/xcodebuild-event-parser.ts [hunk 6]
```diff
@@ -0,0 +1,427 @@
...
+    const pendingFailures = pendingFailureDiagnostics.get(key);
+    if (!pendingFailures) {
+      return;
+    }
+
+    for (const failure of pendingFailures) {
+      emitFailureEvent({ ...failure, durationMs });
+    }
+    pendingFailureDiagnostics.delete(key);
+  }
```
[the remark this answers is not in the document]
Comment (cameroncooke) [reply src/utils/xcodebuild-event-parser.ts:214]: Fixed in 8415cd74. Now deletes entries from `pendingFailureDurations` after consuming them in both `queueFailureDiagnostic` and `applyFailureDuration`.

File: src/utils/swift-testing-line-parsers.ts [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+    return null;
+  }
+
+  const total = Number(match[1]);
+  const durationText = `${match[2]}s`;
+
+  const issueMatch = line.match(/with (\d+) issues?/u);
+  const failed = issueMatch ? Number(issueMatch[1]) : 0;
+
+  return { executed: total, failed, durationText };
```
[the remark this answers is not in the document]
Comment (cameroncooke) [reply src/utils/swift-testing-line-parsers.ts]: Known limitation. Swift Testing's summary reports "issues" not "failed tests" and the framework doesn't provide a distinct failed-test count. Added a clarifying comment in 8415cd74. Downstream reconciliation via `Math.max(failedTests, testFailures.length)` partially mitigates overcounting.

File: src/utils/xcodebuild-event-parser.ts [hunk 8]
```diff
@@ -0,0 +1,427 @@
...
+      return;
+    }
+
+    // Swift Testing continuation line (↳) appends context to pending issue
+    const stContinuation = parseSwiftTestingContinuationLine(line);
+    if (stContinuation) {
+      const lastQueuedEntry = Array.from(pendingFailureDiagnostics.values()).at(-1)?.at(-1);
+      if (lastQueuedEntry) {
+        lastQueuedEntry.message += `\n${stContinuation}`;
+        return;
```
[the remark this answers is not in the document]
Comment (cameroncooke) [reply src/utils/xcodebuild-event-parser.ts:281]: Not a bug. JavaScript `Map` preserves insertion order, so `Array.from(values()).at(-1)` returns the most recently inserted key's array. Combined with `.at(-1)` on that array, this is the most recently queued diagnostic. xcodebuild continuation lines always immediately follow their parent diagnostic, so this ordering is correct.

File: src/utils/xcodebuild-run-state.ts [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+
+export interface RunStateOptions {
+  operation: XcodebuildOperation;
+  minimumStage?: XcodebuildStage;
+  onEvent?: (event: PipelineEvent) => void;
+}
+
+function normalizeDiagnosticKey(location: string | undefined, message: string): string {
+  return `${location ?? ''}|${message}`.trim().toLowerCase();
+}
```
[the remark this answers is not in the document]
Comment (cameroncooke) [reply src/utils/xcodebuild-run-state.ts:35]: Same concern as the earlier Sentry bot comment -- already addressed. The key formats are structurally distinct across event types and can't collide in practice.

File: src/utils/xcodebuild-run-state.ts [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+      for (const tailEvent of options?.tailEvents ?? []) {
+        accept(tailEvent);
+      }
+
+      return { ...state };
+    },
+
+    snapshot(): Readonly<XcodebuildRunState> {
+      return { ...state };
+    },
```
[the remark this answers is not in the document]
Comment (cameroncooke) [reply src/utils/xcodebuild-run-state.ts:251]: Fixed in 1f3cb73b. Both `snapshot()` and `finalize()` now deep-copy all array fields (`events`, `milestones`, `warnings`, `errors`, `testFailures`) so returned state is truly immutable.

File: src/utils/xcodebuild-event-parser.ts [hunk 11]
```diff
@@ -0,0 +1,429 @@
...
+
+const IGNORED_NOISE_PATTERNS = [
+  /^Command line invocation:$/u,
+  /^\s*\/Applications\/Xcode[^\s]+\/Contents\/Developer\/usr\/bin\/xcodebuild\b/u,
+  /^\d{4}-\d{2}-\d{2}\s+\d{2}:\d{2}:\d{2}\.\d+\s+xcodebuild\[.+\]\s+Writing error result bundle to\s+/u,
+  /^Build settings from command line:$/u,
+  /^(?:COMPILER_INDEX_STORE_ENABLE|ONLY_ACTIVE_ARCH)\s*=\s*.+$/u,
+  /^Resolve Package Graph$/u,
+  /^Resolved source packages:$/u,
+  /^\s*[A-Za-z0-9_.-]+:\s+.+$/u,
```
[the remark this answers is not in the document]
Comment (cameroncooke) [reply src/utils/xcodebuild-event-parser.ts]: Fixed in 1f3cb73b. Replaced the broad `/^\s*[A-Za-z0-9_.-]+:\s+.+$/` with `/^\s*[A-Za-z0-9_.-]+:\s+https?:\/\/.+$/` so it only matches SPM resolved dependency lines (the intended target) and no longer swallows compiler `note:` diagnostics.

File: src/utils/swift-testing-event-parser.ts [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+    if (stResult) {
+      completedCount += 1;
+      if (stResult.status === 'failed') {
+        failedCount += 1;
+      }
+      if (stResult.status === 'skipped') {
+        skippedCount += 1;
+      }
+      emitTestProgress();
+      return;
```
[the remark this answers is not in the document]
Comment (cameroncooke) [reply src/utils/swift-testing-event-parser.ts:135]: Fixed in 1f3cb73b. The result line is now checked before flushing the pending issue. When a failed result line arrives with a pending issue diagnostic, the duration from the result line is attached via `parseDurationMs` before emitting the `test-failure` event. Passed/skipped results and failures without a pending issue follow the original path.

File: src/utils/swift-testing-event-parser.ts [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+
+    // Swift Testing result line: ✔/✘/◇ Test "Name" passed/failed/skipped (non-failure or no pending issue)
+    if (stResult) {
+      completedCount += 1;
+      if (stResult.status === 'skipped') {
+        skippedCount += 1;
+      }
+      emitTestProgress();
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (cameroncooke) [reply src/utils/swift-testing-event-parser.ts:136]: Fixed in 5b1e7851. Added the missing `failedCount += 1` for `'failed'` status in the fallthrough result handler. This was a regression from the duration-attachment fix where the early-return path correctly incremented but the fallthrough did not.

File: src/utils/xcodebuild-run-state.ts [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+    ): XcodebuildRunState {
+      state.finalStatus = succeeded ? 'SUCCEEDED' : 'FAILED';
+      state.wallClockDurationMs = durationMs ?? null;
+
+      if (options?.emitSummary !== false) {
+        const reconciledFailedTests = Math.max(state.failedTests, state.testFailures.length);
+        const reconciledPassedTests = Math.max(
+          0,
+          state.completedTests - state.failedTests - state.skippedTests,
+        );
```
[the remark this answers is not in the document]
Comment (cameroncooke) [reply src/utils/xcodebuild-run-state.ts:203]: Intentional. The mismatch between `state.failedTests` and `reconciledFailedTests` represents incomplete information: the progress counter saw fewer failures than the diagnostic parser found. Using `reconciledFailedTests` for the passed calculation would retroactively reclassify a test the counter reported as passed, which is wrong. The total exceeding `completedTests` correctly represents "we know about more failures than the counter saw."

File: src/utils/swift-testing-line-parsers.ts [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+ *   ✘ Test "Name" (aka 'func()') recorded an issue at File.swift:48:5: msg
+ *   ✘ Test "Name" recorded an issue with 1 argument value → 0 at File.swift:10:5: msg
+ *   ✘ Test "Name" recorded an issue: message
+ */
+export function parseSwiftTestingIssueLine(line: string): ParsedFailureDiagnostic | null {
+  // Match with location -- handle both aka suffix and parameterized argument values before "at"
+  const locationRegex = new RegExp(
+    `^[✘] Test "(.+)"${OPTIONAL_AKA} recorded an issue(?:\\s+with \\d+ argument values?[^:]*?)? at (.+?):(\\d+):\\d+: (.+)$`,
+    'u',
+  );
```
[the remark this answers is not in the document]
Comment (cameroncooke) [reply src/utils/swift-testing-line-parsers.ts:96]: Fixed in fdad5b64. Changed `[^:]*?` to `.*?` in the argument value match. The `(.+?):(\d+):\d+:` anchor for the file path is specific enough to backtrack correctly past colons in argument values.

File: src/utils/xcodebuild-event-parser.ts [hunk 16]
```diff
@@ -0,0 +1,429 @@
...
+        testCase.suiteName,
+        testCase.testName,
+        parseDurationMs(testCase.durationText),
+      );
+    }
+    if (testCase.status === 'skipped') {
+      skippedCount += 1;
+    }
+    emitTestProgress();
+  }
```
[the remark this answers is not in the document]
Comment (cameroncooke) [reply src/utils/xcodebuild-event-parser.ts:251]: Fixed in ce10d300. `recordTestCaseResult` now uses `testCase.caseCount ?? 1` for all count increments. Also made the XCTest fallback path in the Swift Testing event parser consistent with the same pattern.

File: src/utils/xcodebuild-event-parser.ts [hunk 17]
```diff
@@ -0,0 +1,431 @@
...
+}
+
+const IGNORED_NOISE_PATTERNS = [
+  /^Command line invocation:$/u,
+  /^\s*\/Applications\/Xcode[^\s]+\/Contents\/Developer\/usr\/bin\/xcodebuild\b/u,
+  /^\d{4}-\d{2}-\d{2}\s+\d{2}:\d{2}:\d{2}\.\d+\s+xcodebuild\[.+\]\s+Writing error result bundle to\s+/u,
+  /^Build settings from command line:$/u,
+  /^(?:COMPILER_INDEX_STORE_ENABLE|ONLY_ACTIVE_ARCH)\s*=\s*.+$/u,
+  /^Resolve Package Graph$/u,
+  /^Resolved source packages:$/u,
```
[the remark this answers is not in the document]
Comment (cameroncooke) [reply src/utils/xcodebuild-event-parser.ts]: Fixed in 7c13ea90. Removed the two dead patterns -- they were already matched by `resolveStageFromLine` via `packageResolutionPatterns` before reaching the noise filter.

File: src/utils/xcodebuild-line-parsers.ts [hunk 18]
```diff
@@ -0,0 +1,174 @@
...
+
+export function parseTotalsLine(line: string): ParsedTotals | null {
+  const match = line.match(
+    /^Executed (\d+) tests?, with (\d+) failures?(?: \(\d+ unexpected\))? in (.+)$/u,
+  );
+  if (!match) {
+    return null;
+  }
+  return { executed: Number(match[1]), failed: Number(match[2]), durationText: match[3] };
+}
```
[the remark this answers is not in the document]
Comment (cameroncooke) [reply src/utils/xcodebuild-line-parsers.ts:108]: Renamed to `displayDurationText` in 7c13ea90 to make the intent explicit. The XCTest totals format (`1.234 (1.235) seconds`) is not parseable by `parseDurationMs` and this field is only used for display.
